src/gadgets/nonnative/bn/big_uint_mul_mod_chip/mod.rs

```diff
@@ -1046,7 +1046,115 @@ impl<F: ff::PrimeField> BigUintMulModChip<F> {
 
         self.is_equal(ctx, grouped_left, grouped_right)?;
 
-        Ok(MultModResult {
+        Ok(ModOperationResult {
+            quotient: assigned_q,
+            remainder: assigned_r,
+        })
+    }
+
+    /// Performs modular reduction of `val` by `modulus`.
+    ///
+    /// This method is part of the Halo2 protocol's arithmetic operations on big integers.
+    /// It reduces a big integer value (`val`) modulo a given modulus (`modulus`) efficiently
+    /// using the `BigUintMulModChip`. The operation is crucial in cryptographic computations
+    /// where maintaining values within a finite field is necessary.
+    ///
+    /// # Arguments
+    /// * `ctx`: mutable reference to the `RegionCtx` which provides the constraint system
+    ///   and metadata necessary for the operation within the Halo2 protocol.
+    /// * `val`: array of `AssignedCell` representing the value to be reduced.
+    /// * `modulus`: array of `AssignedCell` representing the modulus for the reduction.
+    ///
+    /// # Order of Operations
+    /// 1. Convert `val` and `modulus` to `BigUint` objects using
+    ///    [`big_uint::BigUint::from_assigned_cells`].
+    /// 2. Perform the modular reduction `val % modulus` using the converted `BigUint` numbers.
+    /// 2.1. This includes the calculation of the quotient and remainder.
+    /// 3. Assign the reduction result using [`Self::assign_mult`] for `q * m` (quotient times modulus)
+    ///    and [`Self::assign_sum`] for `q * m + r` (sum of quotient times modulus and remainder).
+    /// 4. Group the limbs of the calculated values using [`Self::group_limbs`].
+    /// 5. Check for equivalence between the original value and the calculated sum using
+    ///    [`Self::is_equal`].
+    ///
+    /// # Returns
+    /// * A result wrapping [`ModOperationResult`] object containing the calculated quotient
+    ///   and remainder from the modular reduction.
+    ///
+    /// # Errors
+    /// This method will return an error if there is an issue in the conversion of `val` or
+    /// `modulus` into `BigUint`, or if the assignment of multiplication and addition fails.
+    pub fn red_mod(
+        &self,
+        ctx: &mut RegionCtx<'_, F>,
+        val: &[AssignedCell<F, F>],
+        modulus: &[AssignedCell<F, F>],
+    ) -> Result<ModOperationResult<F>, Error> {
+        // lhs * rhs = q * m + r
+
+        let to_bn = |val| {
+            big_uint::BigUint::from_assigned_cells(val, self.limb_width, self.limbs_count_limit)
```

Comment:
> (1) could you remind me what is "limbs_count_limit" here? what is it typical value?

This is an upper limit on the number of limbs. For the evidence itself, it is always aligned to it, however, the representations of the limbs themselves <= this number

> used to limit the max_word to be all_ones

Yes, just instead of a mathematical constraint, we use a simpler byte constraint. The number of bytes in the field element representation must be less than width